src/adr/0034-distributed-listeners-configuration.md

```diff
@@ -0,0 +1,106 @@
+# 34. Distributed Mailbox Listeners Configuration
+
+Date: 2020-04-23
+
+## Status
+
+Proposed
+
+Supercedes [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md)
+
+## Context
+
+James enables a user to register additional mailbox listeners.
+
+The distributed James server is handling mailbox event processing (mailboxListener execution) using a RabbitMQ work-queue
+per listener.
+
+Currently, mailbox listeners are determined by the guice bindings of the server and additional mailbox listeners defined
+via configuration files.
+
+While the configuration might be specific for each James server, what actually is defined in RabbitMQ is common. 
+Heterogeneous configuration might then result in unpredictable RabbitMQ resource status. This was left as a limitation
+of [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md).
+
+## Decision
+
+We need to centralize the definition of mailbox listeners.
+
+An event sourcing system will track the configured mailbox listeners.
+
+It will have the following commands:
+
+ - **AddListener**: Add a given listener. This should be rejected if the group is already used.
+ - **RemoveListener**: Remove a given listener.
+
+It will have the following events:
+
+ - **ListenerAdded**: A mailbox listener is added
+ - **ListenerRemoved**: A mailbox listener is removed
+
+A subscriber will react to these events to modify the RabbitMQ resource accordingly by adding queues, adding or removing
+bindings.
+
+This event sourcing system differs from the one defined in
+[26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md) by the
+fact that we should also keep track of listener configuration.
+
+Upon start, James will ensure the **configured mailbox listener event sourcing system** contains the guice injected 
+listeners, and add them if missing (handling the RabbitMQ bindings by this mean), then starts the eventBus which will
+consume the given queues.
+
+If a listener is configured with a class unknown to James, the start-up fails. This can happen if a custom jar, containing
+user implemented mailbox listeners is partially deployed.
+
+This differs from [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md)
+by the fact we no longer need to register all listeners at once.
+
+A WebAdmin endpoint will allow:
+
+ - **to add a listener** to the one configured. Such a call:
+    - Will fail if the listener is unknown to James, or if the corresponding group is unknown.
+    - Upon success the listener is added to the **configured mailbox listener aggregate**, and the listener is 
```

Comment:
why?

Reply:
> Upon success the listener is added to the **configured mailbox listener aggregate**

What would you expect else? ^^

Reply:
Sorry, I failed to write a proper comment. Let me try again:

> No broadcast is attempted, meaning that other James servers will need a reboot to actually start 
    consuming the queue.

Why?

Reply:
Because doing a broadcast requires more dev to take place. 

This is listed as a possible enhancement below.

Reply:
> Because doing a broadcast requires more dev to take place.

Do you think so? Because I'm really not sure regarding what we will have to do to deal with the "no broadcast" situation.

BTW, we already have an EventBus ...

Reply:
> Do you think so? Because I'm really not sure regarding what we will have to do to deal with the "no broadcast" situation.

Let's move this to a TLD concept then.

Reply:
> BTW, we already have an EventBus ...

So. Via a Registration key, to leverage a broadcast? Smart.